lib/Onyx.js

```diff
@@ -590,6 +643,15 @@ function sendDataToConnection(mapping, val, matchedKey) {
             }
         }
 
+        // sendDataToConnection only get's called in the connect code.
+        // withOnyx is already setting data from cache initially. If we now
+        // want to call sendDataToConnection, we want to make sure, we
+        // are not sending again the same data from cache, if the component
+        // already has the cached data.
+        if (deepEqual(mapping.withOnyxInstance.state[mapping.statePropertyName], newData)) {
+            return;
+        }
```

Comment:
Do we need a `deepEqual()` here? Wouldn't it be enough to see that it has the property name at all? It must have been set in the constructor? Or maybe I am missing some situation where these could actually be different.

Reply:
Hm, you are right in a way. I think deepEqual is overkill, and a simple ref comparison should be enough. However we should do a check here. There is the situation where a prop provided to the component using withOnyx changes, which changes the key requested. In this case we connect the component again, and this code gets called again. However, at this point in time there will already be a state value. Thats why we can't return simply when the value seems to be in state>

Reply:
ah lol no, that wouldn't be enough, because we also have selectors, and their return value is always a different ref, even if the input data have been the same. Hmmm ...

Reply:
Okay, so i optimised it to use deepEqual for selectors only. Please check again

Reply:
>prop provided to the component using withOnyx changes, which changes the key requested. In this case we connect the component again, and this code gets called again. However, at this point in time there will already be a state value. Thats why we can't return simply when the value seems to be in state.

Whoaaa! Did we see this happen? 🙃 

Let me make sure I understand the order of operations:

- Connect to a key with a prop (`withOnyx()` `constructor()` called)
- Component mounts and there is no value for something in cache so we try get the stored value (async operation)
- _Maybe_ we changed the prop influencing this key while we are getting the data from storage
- Component updates and we `disconnect()` (but don't cancel the reading from storage that is happening)
- We finish reading the first thing from storage and call `sendDataToConnection()`
- It will return early due to the logic on line 629

**Question:** What happens to the `loading` state in this scenario since we are never calling `setWithOnyxState()` and the value of `requiredKeysForInit` is never updated when the other key is disconnected? Does the component never render?

FWIW I am still not getting why we need the `deepEqual()`. Maybe clearer way to look at this is...

If we have the cached data then why are we calling `sendDataToConnection()` at all? Maybe we pass something to [this method](https://github.com/Expensify/react-native-onyx/blob/d4fc53cc3229808e9c1c542c13ce0089e5893d96/lib/withOnyx.js#L146C13-L157) and just skip it during the `connect()` method.